hal/src/rtl872x/ble_hal.cpp

```diff
@@ -1616,8 +1681,13 @@ int BleGap::startScanning(hal_ble_on_scan_result_cb_t callback, void* context) {
     CHECK_RTL(le_scan_start());
     isScanning_ = true;
     // GAP_SCAN_STATE_SCANNING may be propagated immediately following the GAP_SCAN_STATE_START
-    if (waitState(BleGapDevState().scan(GAP_SCAN_STATE_SCANNING))) {
-        return SYSTEM_ERROR_TIMEOUT;
+    if (waitState(BleGapDevState().scan(GAP_SCAN_STATE_START), 10, true) && waitState(BleGapDevState().scan(GAP_SCAN_STATE_SCANNING), 10, true)) {
```

Comment:
10ms sounds like a very tight condition?

Reply:
That's desinated. I just utilize the `waitState()` to do a quick check before a long wait.

Reply:
>Is it critical to call BleGapDevState().scan(GAP_SCAN_STATE_START) before BleGapDevState().scan(GAP_SCAN_STATE_SCANNING)?

Nope. Just treat it as a check (`force` poll is set to `true`) without waiting any time (although there is 10ms). The `GAP_SCAN_STATE_START` and `GAP_SCAN_STATE_SCANNING` sometimes are notified in wrong sequence.